### ISSUE TITLE
Officialy support newer versions of PostgreSQL

### ISSUE THREAD

**Issue by semangard** [keycloak/keycloak#10767]
### Description

Based on the [documentation](https://www.keycloak.org/server/db):
![image](https://user-images.githubusercontent.com/57065741/158589604-bb87a0cc-6c28-47eb-953c-27bab8e59107.png)
Only the version 10 of PostgreSQL is officialy supported.

This is quite an old version and even the associated JDBC driver seems also very old 
https://github.com/keycloak/keycloak-containers/blob/main/server/Dockerfile (for KC legacy)
vs
https://jdbc.postgresql.org/documentation/changelog.html#version_all

==> Would it be possible to suppport newer versions of PostgreSQL (especially if PostgreSQL is one of the 2 main DB supported by KeyCloak) ?

### Motivation

Benefit of latest enhancements and fxxes of PostgreSQL
and follow up supported versions of PosgtreSQL 

https://www.postgresql.org/support/versioning/
![image](https://user-images.githubusercontent.com/57065741/158590145-ceefde69-8250-48ee-86bc-d7529a8cee17.png)

### Discussion
https://github.com/keycloak/keycloak/discussions/9232

**Comment by myfirstnameispaul** [comment]
Your link seems to be for the Keycloak legacy repo?

https://github.com/keycloak/keycloak/blob/99f27497f482afbdd281a90dacdc2dc6f24e97ec/quarkus/pom.xml#L45

**Comment by stianst** [comment]
With the legacy store and the number of databases supported we're not able to test multiple versions of each database. We are looking at updating the versions we are testing with though.

For the new store and PostgreSQL the plan is to test with additional versions, including updating what version we are testing with much faster when new releases are available for PostgreSQL.

**Comment by radiodup** [comment]
How could we do/execute test in order to test and share results over  postgres 14 support?
Is there any suite? tools? guide?

**Comment by qhaas** [comment]
As of 11/24/22, the [documentation](https://www.keycloak.org/server/db#_supported_databases) still shows postgres 10 as the supported version, postgres 10 [is now EOL](https://www.postgresql.org/support/versioning).

**Comment by Deltahedron** [comment]
This is quite weird from configuration side, some pointers on what to test as @qhaas mentioned would be appreciated!

**Comment by Siedlerchr** [comment]
Still not any udpate on the doc side, anyone already got it running with a newer verison as 10?

**Comment by RonneGisun** [comment]
Amazon is forcing an upgrade from the end of this month as Postgres 10 will be out of support.
https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-release-calendar.html
So it would be nice to see an official announcement as to newer supported Postgres versions.

I have had Keycloak version 15 successfully running with Postgres 12.7 for some time now.
I am currently testing Keycloak legacy version 19, that runs on Wildfly 25, with Postgres 13.7.
I would be interested to hear if anyone has tested any version of Keycloak with Postgres 14?

**Comment by Siedlerchr** [comment]
We have latest Keycloack running with Postgres 12.x on AWS

**Comment by radiodup** [comment]
Oka. No an official answer so far.

But I dug a little and I've found:

https://github.com/keycloak/keycloak/blob/main/docs/tests-db.md

So, there is a test suit to run against several engines.

I think that we can choose a version (I'll do wildfly 16.1.1 with postgresql 14.6) and post results here or anywhere.

We might use or reuse a wiki? modify documents?

And this will serve the community at least in an amateur level (not a developer myself)

**Comment by radiodup** [comment]
### Caveat emptor: "CRASH COURSE ON TESTING FOR PEOPLE THAT DON'T WANT TO KNOW, JUST DO"

Anyway, you have to read at least this:

https://github.com/keycloak/keycloak/blob/16.1.1/docs/tests.md
https://github.com/keycloak/keycloak/blob/16.1.1/docs/building.md


Validate this:
```
mvn -version
Apache Maven 3.6X
Maven home: /X/maven
Java version: 11.0.# vendor: X /X/java-11-openjdk-#
Default locale: es_AR, platform encoding: UTF-8
OS name: "linux", version: "X", arch: "amd64", family: "unix"

java -version
openjdk version "11.0.X
OpenJDK Runtime Environment (X11.0.X) (build 11.0X)
OpenJDK 64-Bit Server VM (X) (build X)

git --version
git version 2.3X
```

Download the project
```
cd projects
git clone https://github.com/keycloak/keycloak.git
cd keycloak 
git checkout tags/16.1.1
```

Edit the pom.xml file to add a property to maven-surefire-plugin. The plugin is already configured, but the loggin manager is not configured (something probably done by default by people who knows/does testing)
```
vi pom.xml

#<plugin>
#	<artifactId>maven-surefire-plugin</artifactId>
#	<version>2.12</version>
#	<configuration>
#    	<systemProperties>
#        	<property>
#            	<name>java.util.logging.manager</name>
#            	<value>org.jboss.logmanager.LogManager</value>
#        	</property>
#    	</systemProperties>
#	</configuration>
#</plugin>
```

If your dns provider filter dns request that resolves to private ip adresses, change it to 8.8.8.8 or something that allows to resolve: (this is on the testing doc)
```
keycloak.127.0.0.1.nip.io
proxy.kc.127.0.0.1.nip.io
localtest.me

```

They should resolve to 127.0.0.1 (.nip.io) and ::1 (.me)

### Run the test 
(might be avoidable, but just in case, i've run the test with the h2, and then did it with postgres)

```
mvn clean install
```


### Run the database test

https://github.com/keycloak/keycloak/blob/16.1.1/docs/tests-db.md

Create a docker with postgres to run the tests and run the test pointing to the recently created database, I've done it with a postgres 14.6 oficial image.

```docker run --name postgres -e POSTGRES_DATABASE=keycloak -e POSTGRES_USER=keycloak -e POSTGRES_PASSWORD=keycloak -e POSTGRES_ROOT_PASSWORD=keycloak -d postgres:14.6
```

```
mvn install -Dkeycloak.connectionsJpa.url=jdbc:postgresql://`docker inspect --format '{{ .NetworkSettings.IPAddress }}' postgres`:5432/keycloak -Dkeycloak.connectionsJpa.driver=org.postgresql.Driver -Dkeycloak.connectionsJpa.user=keycloak -Dkeycloak.connectionsJpa.password=keycloak  
```


```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Keycloak 999.0.0-SNAPSHOT:
[INFO] 
[INFO] Keycloak BOM Parent ................................ SUCCESS [  0.691 s]
[INFO] Keycloak BOM for adapters .......................... SUCCESS [  0.008 s]
[INFO] Keycloak BOM for server extensions ................. SUCCESS [  0.007 s]
[INFO] Keycloak BOM utilities for the quickstarts ......... SUCCESS [  0.013 s]
[INFO] Keycloak ........................................... SUCCESS [  0.031 s]
[INFO] Keycloak Common .................................... SUCCESS [  2.531 s]
[INFO] Keycloak Core ...................................... SUCCESS [  1.554 s]
[INFO] Keycloak Server SPI ................................ SUCCESS [  1.060 s]
[INFO] Keycloak Server Private SPI ........................ SUCCESS [  4.237 s]
[INFO] Keycloak Crypto Parent ............................. SUCCESS [  0.011 s]
[INFO] Keycloak Crypto Default ............................ SUCCESS [  4.057 s]
[INFO] Keycloak SAML Core Public API ...................... SUCCESS [  0.788 s]
[INFO] Keycloak SAML Core ................................. SUCCESS [  2.705 s]
[INFO] Keycloak REST Services ............................. SUCCESS [  5.341 s]
[INFO] Keycloak Crypto FIPS 140-2 Integration ............. SUCCESS [  3.819 s]
[INFO] Keycloak Crypto Wildfly Elytron Integration ........ SUCCESS [  2.304 s]
[INFO] Keycloak JavaScript Adapter (JAR) .................. SUCCESS [ 20.474 s]
[INFO] Keycloak Themes .................................... SUCCESS [ 51.436 s]
[INFO] Keycloak Dependencies Parent ....................... SUCCESS [  0.009 s]
[INFO] Keycloak Dependencies Server Min ................... SUCCESS [  0.015 s]
[INFO] Keycloak Model Parent .............................. SUCCESS [  0.012 s]
[INFO] Keycloak Legacy Database Support ................... SUCCESS [  0.118 s]
[INFO] Keycloak Legacy Database Support - Private classes . SUCCESS [  0.544 s]
[INFO] Keycloak Model JPA ................................. SUCCESS [  0.741 s]
[INFO] Keycloak Model Java Annotations and Processor ...... SUCCESS [  0.043 s]
[INFO] Keycloak Model Map ................................. SUCCESS [  1.092 s]
[INFO] Keycloak Model Map JPA ............................. SUCCESS [  0.784 s]
[INFO] Keycloak Model Map LDAP ............................ SUCCESS [  0.330 s]
[INFO] Keycloak Kerberos Federation ....................... SUCCESS [  0.023 s]
[INFO] Keycloak LDAP UserStoreProvider .................... SUCCESS [  0.479 s]
[INFO] Keycloak Legacy Database Support - REST services ... SUCCESS [  0.051 s]
[INFO] Keycloak Model Infinispan .......................... SUCCESS [  1.933 s]
[INFO] Keycloak Model Hot Rod ............................. SUCCESS [  0.781 s]
[INFO] Keycloak Model Map File ............................ SUCCESS [  0.031 s]
[INFO] Keycloak SSSD Federation ........................... SUCCESS [  0.194 s]
[INFO] Keycloak Authz: Parent ............................. SUCCESS [  0.006 s]
[INFO] Keycloak AuthZ: Policy Provider Parent ............. SUCCESS [  0.005 s]
[INFO] Keycloak AuthZ: Common Policy Providers ............ SUCCESS [  0.048 s]
[INFO] Keycloak Dependencies Server All ................... SUCCESS [  0.016 s]
[INFO] Keycloak Quarkus Parent ............................ SUCCESS [  0.007 s]
[INFO] Keycloak Configuration API ......................... SUCCESS [  0.047 s]
[INFO] Keycloak JavaScript Parent ......................... SUCCESS [ 18.523 s]
[INFO] Keycloak Admin UI .................................. SUCCESS [ 42.427 s]
[INFO] Keycloak Administration UI ......................... SUCCESS [  0.005 s]
[INFO] Admin UI REST extensions ........................... SUCCESS [  2.098 s]
[INFO] Keycloak Quarkus Server Extension .................. SUCCESS [  4.051 s]
[INFO] Keycloak Docs Parent ............................... SUCCESS [  0.005 s]
[INFO] Keycloak Guides Maven Plugin ....................... SUCCESS [  3.788 s]
[INFO] Keycloak Guides .................................... SUCCESS [  8.305 s]
[INFO] Keycloak Federation ................................ SUCCESS [  0.007 s]
[INFO] Keycloak Util Embedded LDAP ........................ SUCCESS [  0.133 s]
[INFO] Keycloak Util Parent ............................... SUCCESS [  0.007 s]
[INFO] Keycloak Integration ............................... SUCCESS [  0.006 s]
[INFO] Keycloak Admin REST Client ......................... SUCCESS [  0.086 s]
[INFO] Keycloak Admin REST Client for Jakarta packages .... SUCCESS [  0.792 s]
[INFO] Keycloak Client Registration API ................... SUCCESS [  0.023 s]
[INFO] Keycloak Client CLI ................................ SUCCESS [  0.005 s]
[INFO] Keycloak Client Registration CLI ................... SUCCESS [  1.089 s]
[INFO] Keycloak Admin CLI ................................. SUCCESS [  1.032 s]
[INFO] Keycloak Client CLI Distribution ................... SUCCESS [  0.379 s]
[INFO] Keycloak Adapter SPI ............................... SUCCESS [  0.128 s]
[INFO] Keycloak Tomcat Adapter SPI ........................ SUCCESS [  0.046 s]
[INFO] Keycloak Undertow Integration SPI .................. SUCCESS [  0.144 s]
[INFO] Keycloak Servlet Integration ....................... SUCCESS [  0.101 s]
[INFO] Common JBoss/Wildfly Core Classes .................. SUCCESS [  0.028 s]
[INFO] Keycloak Jetty Adapter SPI ......................... SUCCESS [  0.159 s]
[INFO] Keycloak Client Adapter SPI Modules ................ SUCCESS [  0.006 s]
[INFO] Keycloak SAML Client Adapter Public API ............ SUCCESS [  0.028 s]
[INFO] Keycloak SAML Client Adapter Core .................. SUCCESS [  1.442 s]
[INFO] Keycloak Jetty Core SAML Integration ............... SUCCESS [  0.121 s]
[INFO] Keycloak Jetty 9.4.x SAML Integration .............. SUCCESS [  0.145 s]
[INFO] Keycloak SAML Jetty Integration .................... SUCCESS [  0.005 s]
[INFO] Keycloak Undertow SAML Adapter ..................... SUCCESS [  0.036 s]
[INFO] Keycloak SAML Tomcat Integration ................... SUCCESS [  0.005 s]
[INFO] Keycloak Tomcat Core SAML Integration .............. SUCCESS [  0.033 s]
[INFO] Keycloak Tomcat SAML Integration ................... SUCCESS [  0.032 s]
[INFO] Keycloak Authz: Client API ......................... SUCCESS [  0.129 s]
[INFO] Keycloak Authz: Policy Enforcer .................... SUCCESS [  0.056 s]
[INFO] Keycloak Adapter Core .............................. SUCCESS [  1.531 s]
[INFO] Keycloak WildFly Elytron SAML Adapter .............. SUCCESS [  0.055 s]
[INFO] Keycloak Wildfly SAML Adapter Subsystem ............ SUCCESS [  2.390 s]
[INFO] Keycloak SAML Wildfly Integration .................. SUCCESS [  0.004 s]
[INFO] Keycloak SAML Servlet Filter ....................... SUCCESS [  0.018 s]
[INFO] Keycloak SAML Client Adapter Modules ............... SUCCESS [  0.004 s]
[INFO] Keycloak Installed Application ..................... SUCCESS [  0.019 s]
[INFO] Keycloak JAX-RS OAuth Client ....................... SUCCESS [  0.021 s]
[INFO] Keycloak Jetty Core Integration .................... SUCCESS [  0.095 s]
[INFO] Keycloak Jetty 9.4.x Integration ................... SUCCESS [  0.092 s]
[INFO] Keycloak Jetty Integration ......................... SUCCESS [  0.006 s]
[INFO] Keycloak Servlet Filter Adapter Integration ........ SUCCESS [  0.099 s]
[INFO] Keycloak Tomcat Integration ........................ SUCCESS [  0.007 s]
[INFO] Keycloak Tomcat Core Integration ................... SUCCESS [  0.029 s]
[INFO] Keycloak Tomcat Integration ........................ SUCCESS [  0.027 s]
[INFO] Keycloak Undertow Integration ...................... SUCCESS [  0.170 s]
[INFO] spring-boot-container-bundle ....................... SUCCESS [  0.079 s]
[INFO] Keycloak Spring Security Integration ............... SUCCESS [  2.019 s]
[INFO] Keycloak Spring Boot Adapter Core .................. SUCCESS [  0.118 s]
[INFO] Keycloak Spring Boot 2 Integration ................. SUCCESS [  0.772 s]
[INFO] Keycloak Wildfly Elytron OIDC Adapter .............. SUCCESS [  0.028 s]
[INFO] Keycloak Wildfly Adapter Subsystem ................. SUCCESS [  1.946 s]
[INFO] Keycloak WildFly Integration ....................... SUCCESS [  0.004 s]
[INFO] Keycloak OIDC Client Adapter Modules ............... SUCCESS [  0.003 s]
[INFO] Keycloak Adapters .................................. SUCCESS [  0.004 s]
[INFO] Keycloak Misc ...................................... SUCCESS [  0.003 s]
[INFO] Keycloak :: Spring :: Boot ......................... SUCCESS [  0.003 s]
[INFO] Keycloak :: Spring :: Boot :: Default ::  Starter .. SUCCESS [  0.026 s]
[INFO] keycloak-test-helper ............................... SUCCESS [  0.058 s]
[INFO] Keycloak JavaScript Admin Client ................... SUCCESS [ 19.997 s]
[INFO] Keycloak JavaScript Adapter ........................ SUCCESS [ 19.187 s]
[INFO] Keycloak Quarkus Server Deployment ................. SUCCESS [ 41.105 s]
[INFO] Keycloak Quarkus Server App ........................ SUCCESS [  5.940 s]
[INFO] Keycloak Quarkus Server Distribution ............... SUCCESS [ 17.405 s]
[INFO] Keycloak Quarkus Test Parent ....................... SUCCESS [  0.004 s]
[INFO] Keycloak Quarkus Server Integration tests .......... SUCCESS [28:54 min]
[INFO] Keycloak TestSuite ................................. SUCCESS [  0.006 s]
[INFO] DB Allocator Plugin ................................ SUCCESS [  1.207 s]
[INFO] Keycloak Arquillian Integration TestSuite .......... SUCCESS [  0.005 s]
[INFO] Test apps .......................................... SUCCESS [  0.005 s]
[INFO] Test apps distribution ............................. SUCCESS [  0.327 s]
[INFO] Keycloak Authz: PhotoZ  Test Parent ................ SUCCESS [  0.004 s]
[INFO] Keycloak Authz Test: Photoz RESTful API ............ SUCCESS [  0.196 s]
[INFO] Keycloak Authz Tests: Photoz HTML5 Client .......... SUCCESS [  0.304 s]
[INFO] Keycloak Authz Tests: Hello World Example .......... SUCCESS [  0.049 s]
[INFO] Keycloak Authz: Servlet Authorization Test ......... SUCCESS [  0.046 s]
[INFO] Keycloak Authz: Simple Servlet App with Policy Enforcer SUCCESS [  0.033 s]
[INFO] integration-arquillian-test-apps-servlets .......... SUCCESS [  0.072 s]
[INFO] Keycloak Test App Profile JEE ...................... SUCCESS [  0.033 s]
[INFO] integration-arquillian-test-apps-cors-parent ....... SUCCESS [  0.005 s]
[INFO] Angular Product Portal JS .......................... SUCCESS [  0.100 s]
[INFO] JAX-RS Database Service Using OAuth Bearer Tokens .. SUCCESS [  0.039 s]
[INFO] spring-boot-adapter-app ............................ SUCCESS [  0.103 s]
[INFO] Servers ............................................ SUCCESS [  0.006 s]
[INFO] Auth Server ........................................ SUCCESS [  0.006 s]
[INFO] Auth Server Services ............................... SUCCESS [  0.007 s]
[INFO] Auth Server Services - Testsuite Providers Deployment SUCCESS [  0.024 s]
[INFO] Auth Server Services - Testsuite Providers ......... SUCCESS [  0.263 s]
[INFO] Keycloak TestSuite Utils ........................... SUCCESS [  0.196 s]
[INFO] Test Util .......................................... SUCCESS [  0.100 s]
[INFO] Auth Server - Undertow ............................. SUCCESS [  0.141 s]
[INFO] Auth Server - Quarkus .............................. SUCCESS [ 11.461 s]
[INFO] App Server ......................................... SUCCESS [  0.007 s]
[INFO] App Server - SPI ................................... SUCCESS [  0.030 s]
[INFO] App Server - JBoss ................................. SUCCESS [  0.007 s]
[INFO] App Server - Tomcat ................................ SUCCESS [  0.007 s]
[INFO] App Server - Undertow .............................. SUCCESS [  0.068 s]
[INFO] App Server - Jetty Parent .......................... SUCCESS [  0.005 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server SUCCESS [  0.004 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server - Infinispan SUCCESS [  0.005 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server - Legacy SUCCESS [  0.005 s]
[INFO] Tests .............................................. SUCCESS [  0.098 s]
[INFO] Base TestSuite ..................................... SUCCESS [  01:55 h]
[INFO] Other Tests Modules ................................ SUCCESS [  0.005 s]
[INFO] Tests for logical storage layer .................... SUCCESS [  2.240 s]
[INFO] Keycloak Examples .................................. SUCCESS [  0.005 s]
[INFO] Keycloak Examples - Admin Client ................... SUCCESS [  0.248 s]
[INFO] Provider Examples .................................. SUCCESS [  0.005 s]
[INFO] REST Example ....................................... SUCCESS [  0.033 s]
[INFO] Domain Extension Example ........................... SUCCESS [  0.061 s]
[INFO] Authenticator Example .............................. SUCCESS [  0.035 s]
[INFO] JS Console ......................................... SUCCESS [  0.020 s]
[INFO] Keycloak Examples - Kerberos Credential Delegation . SUCCESS [  0.061 s]
[INFO] Themes Examples .................................... SUCCESS [  0.031 s]
[INFO] SAML Examples ...................................... SUCCESS [  0.003 s]
[INFO] Keycloak SAML Adapter as a Servlet Filter .......... SUCCESS [  0.404 s]
[INFO] LDAP Demo Application .............................. SUCCESS [  0.037 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:29 h
[INFO] Finished at: 2023-04-07T18:38:51-03:00
[INFO] ------------------------------------------------------------------------

```

keycloak 16.1.1 on postgresql 14.6

Next step would be run all the database operation tests with a non-owner database user (to increase security). But that requires knowledge.

**Comment by ahus1** [comment]
This has been fixed for the upcoming version Keycloak 22: The current main branch is tested with PostgreSQL 15. The version we test with is documented in the root pom.xml file, and will also be published in the docs once they have been released: 

https://github.com/keycloak/keycloak/blob/main/pom.xml#L139-L149

What remains is that we still test only one version of PostgreSQL with Keycloak. If someone wants to test a different version, they can fork this repository, change the version in the pom.xml and run the tests on their own environment or via GitHub actions. 

The next release will show something like the following: 

![image](https://github.com/keycloak/keycloak/assets/3957921/5b1dc665-948d-4e5c-a169-f45bbda83cd4)
